File: frontend/src/services/blockingService.ts
// App Blocking Service with Backend API Integration
import { apiCall } from '../config/api';

export interface BlockedApp {
  id: number;
  name: string;
  category: string;
  icon: string;
  logo?: string;
  blocked: boolean;
  packageName?: string;
  bundleId?: string;
}

export interface BlockingSession {
  id: string;
  userId: string;
  appId: number;
  appName: string;
  startTime: Date;
  duration: number; // in minutes
  status: 'active' | 'completed' | 'broken';
  coinsEarned: number;
  coinsLost: number;
  completedAt?: Date;
  brokenAt?: Date;
}

export interface BlockingLog {
  id: string;
  userId: string;
  appId: number;
  appName: string;
  action: 'block' | 'unblock' | 'session_start' | 'session_complete' | 'session_broken';
  timestamp: Date;
  coinsChange: number;
  sessionId?: string;
}

const REWARD_COINS = 20; // Coins earned for completing session
const PENALTY_COINS = 50; // Coins lost for breaking session
const DEFAULT_SESSION_DURATION = 50; // minutes

class BlockingService {
  /**
   * Get user's blocked apps from backend
   */
  async getUserBlockedApps(userId: string): Promise<BlockedApp[]> {
    try {
      const response = await apiCall('/blocking/apps', 'GET');
      
      return response.map((app: any) => ({
        id: app.appId,
        name: app.appName,
        category: app.category,
        icon: app.icon,
        logo: app.logo,
        blocked: app.blocked,
        packageName: app.packageName,
        bundleId: app.bundleId,
      }));
    } catch (error) {
      console.error('Error getting blocked apps:', error);
      return [];
    }
  }

  /**
   * Save blocked apps to backend
   */
  async saveUserBlockedApps(userId: string, apps: BlockedApp[]): Promise<void> {
    try {
      await apiCall('/blocking/apps', 'POST', {
        apps: apps.map(app => ({
          appId: app.id,
          appName: app.name,
          category: app.category,
          icon: app.icon,
          logo: app.logo,
          blocked: app.blocked,
          packageName: app.packageName,
          bundleId: app.bundleId,
        })),
      });
    } catch (error) {
      console.error('Error saving blocked apps:', error);
      throw error;
    }
  }

  /**
   * Toggle app blocking status
   */
  async toggleAppBlocking(userId: string, appId: number, apps: BlockedApp[]): Promise<BlockedApp[]> {
    try {
      await apiCall(`/blocking/apps/${appId}/toggle`, 'PUT');
      return await this.getUserBlockedApps(userId);
    } catch (error) {
      console.error('Error toggling app blocking:', error);
      throw error;
    }
  }

  /**
   * Start a blocking session for an app
   */
  async startBlockingSession(
    userId: string,
    appId: number,
    appName: string,
    duration: number = DEFAULT_SESSION_DURATION
  ): Promise<BlockingSession> {
    try {
      const response = await apiCall('/blocking/sessions/start', 'POST', {
        appId,
        appName,
        duration,
      });

      return {
        id: response.id,
        userId,
        appId: response.appId,
        appName: response.appName,
        startTime: new Date(response.startTime),
        duration: response.duration,
        status: response.status,
        coinsEarned: response.coinsEarned,
        coinsLost: response.coinsLost,
      };
    } catch (error) {
      console.error('Error starting blocking session:', error);
      throw error;
    }
  }

  /**
   * Complete a blocking session (user successfully avoided the app)
   */
  async completeBlockingSession(sessionId: string, userId: string): Promise<number> {
    try {
      const response = await apiCall(`/blocking/sessions/${sessionId}/complete`, 'POST');
      return response.coinsChange;
    } catch (error) {
      console.error('Error completing blocking session:', error);
      throw error;
    }
  }

  /**
   * Break a blocking session (user unlocked the app mid-session)
   */
  async breakBlockingSession(sessionId: string, userId: string): Promise<number> {
    try {
      const response = await apiCall(`/blocking/sessions/${sessionId}/break`, 'POST');
      return Math.abs(response.coinsChange);
    } catch (error) {
      console.error('Error breaking blocking session:', error);
      throw error;
    }
  }

  /**
   * Get active blocking sessions for user
   */
  async getActiveBlockingSessions(userId: string): Promise<BlockingSession[]> {
    try {
      const response = await apiCall('/blocking/sessions/active', 'GET');
      
      return response.map((s: any) => ({
        id: s.id,
        userId: s.userId,
        appId: s.appId,
        appName: s.appName,
        startTime: new Date(s.startTime),
        duration: s.duration,
        status: s.status,
        coinsEarned: s.coinsEarned,
        coinsLost: s.coinsLost,
      }));
    } catch (error) {
      console.error('Error getting active sessions:', error);
      return [];
    }
  }

  /**
   * Check if app has an active blocking session
   */
  async hasActiveSession(userId: string, appId: number): Promise<BlockingSession | null> {
    try {
      const sessions = await this.getActiveBlockingSessions(userId);
      return sessions.find(s => s.appId === appId) || null;
    } catch (error) {
      console.error('Error checking active session:', error);
      return null;
    }
  }

  /**
   * Get blocking statistics for user
   */
  async getBlockingStats(userId: string): Promise<{
    totalSessions: number;
    completedSessions: number;
    brokenSessions: number;
    totalCoinsEarned: number;
    totalCoinsLost: number;
    successRate: number;
  }> {
    try {
      const response = await apiCall('/blocking/stats', 'GET');
      
      return {
        totalSessions: response.totalSessions,
        completedSessions: response.completedSessions,
        brokenSessions: response.brokenSessions,
        totalCoinsEarned: response.totalCoinsEarned,
        totalCoinsLost: response.totalCoinsLost,
        successRate: response.successRate,
      };
    } catch (error) {
      console.error('Error getting blocking stats:', error);
      return {
        totalSessions: 0,
        completedSessions: 0,
        brokenSessions: 0,
        totalCoinsEarned: 0,
        totalCoinsLost: 0,
        successRate: 0,
      };
    }
  }

  /**
   * Clean up old/stale active sessions (for debugging)
   */
  async cleanupStaleSessions(userId: string): Promise<number> {
    try {
      const response = await apiCall('/blocking/sessions/cleanup', 'POST');
      return response;
    } catch (error) {
      console.error('Error cleaning up stale sessions:', error);
      return 0;
    }
  }

  // Placeholder methods for backward compatibility
  private async updateUserCoins(userId: string, coinsChange: number): Promise<void> {
    // Coins are updated on the backend
    console.log('Coins update handled by backend');
  }

  private async logBlockingAction(
    userId: string,
    appId: number,
    appName: string,
    action: BlockingLog['action'],
    coinsChange: number,
    sessionId?: string
  ): Promise<void> {
    // Logging is handled by backend
    console.log('Logging handled by backend');
  }

  private async getLocalBlockedApps(): Promise<BlockedApp[]> {
    return [];
  }
}

export const blockingService = new BlockingService();
export { DEFAULT_SESSION_DURATION, PENALTY_COINS, REWARD_COINS };

File: backend/src/main/java/com/sankalai/service/BlockingService.java
package com.sankalai.service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.sankalai.dto.BlockingDTO.BlockedAppDTO;
import com.sankalai.dto.BlockingDTO.BlockingSessionDTO;
import com.sankalai.dto.BlockingDTO.BlockingStatsDTO;
import com.sankalai.dto.BlockingDTO.SaveBlockedAppsRequest;
import com.sankalai.dto.BlockingDTO.SessionActionResponse;
import com.sankalai.dto.BlockingDTO.StartSessionRequest;
import com.sankalai.entity.BlockedApp;
import com.sankalai.entity.BlockingLog;
import com.sankalai.entity.BlockingSession;
import com.sankalai.entity.User;
import com.sankalai.entity.UserStats;
import com.sankalai.exception.BadRequestException;
import com.sankalai.exception.ResourceNotFoundException;
import com.sankalai.exception.UnauthorizedActionException;
import com.sankalai.repository.BlockedAppRepository;
import com.sankalai.repository.BlockingLogRepository;
import com.sankalai.repository.BlockingSessionRepository;
import com.sankalai.repository.UserRepository;
import com.sankalai.repository.UserStatsRepository;

@Service
public class BlockingService {

    private static final int REWARD_COINS_PER_BLOCKED_APP = 30;
    private static final int PENALTY_COINS = 50;
    private static final int DEFAULT_SESSION_DURATION = 50;
    private final UserRepository userRepository;
    private final BlockedAppRepository blockedAppRepository;

    private final BlockingSessionRepository blockingSessionRepository;
    private final BlockingLogRepository blockingLogRepository;
    private final UserStatsRepository userStatsRepository;

    public BlockingService(UserRepository userRepository, BlockedAppRepository blockedAppRepository,
            BlockingSessionRepository blockingSessionRepository, BlockingLogRepository blockingLogRepository,
            UserStatsRepository userStatsRepository) {
        this.userRepository = userRepository;
        this.blockedAppRepository = blockedAppRepository;
        this.blockingSessionRepository = blockingSessionRepository;
        this.blockingLogRepository = blockingLogRepository;
        this.userStatsRepository = userStatsRepository;
    }

    /**
     * Get user's blocked apps
     */
    public List<BlockedAppDTO> getUserBlockedApps(String userId) {
        List<BlockedApp> apps = blockedAppRepository.findByUser_UserId(userId);
        return apps.stream()
                .map(this::mapToDTO)
                .collect(Collectors.toList());
    }

    /**
     * Save/Update blocked apps for user
     */
    @Transactional
    public List<BlockedAppDTO> saveUserBlockedApps(String userId, SaveBlockedAppsRequest request) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        // Delete existing blocked apps
        List<BlockedApp> existingApps = blockedAppRepository.findByUser_UserId(userId);
        blockedAppRepository.deleteAll(existingApps);

        // Save new blocked apps
        List<BlockedApp> newApps = request.getApps().stream()
                .map(appReq -> new BlockedApp(user, appReq.getAppId(), appReq.getAppName(), appReq.getCategory(), appReq.getIcon(), appReq.getLogo(), appReq.getBlocked(), appReq.getPackageName(), appReq.getBundleId()))
                .collect(Collectors.toList());

        List<BlockedApp> savedApps = blockedAppRepository.saveAll(newApps);

        return savedApps.stream()
                .map(this::mapToDTO)
                .collect(Collectors.toList());
    }

    /**
     * Toggle app blocking status
     */
    @Transactional
    public BlockedAppDTO toggleAppBlocking(String userId, Integer appId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        BlockedApp app = blockedAppRepository.findByUser_UserIdAndAppId(userId, appId)
                .orElseThrow(() -> new ResourceNotFoundException("App not found"));

        app.setBlocked(!app.getBlocked());
        BlockedApp savedApp = blockedAppRepository.save(app);

        // Log the action
        logBlockingAction(user, appId, app.getAppName(),
                app.getBlocked() ? BlockingLog.LogAction.BLOCK : BlockingLog.LogAction.UNBLOCK,
                0, null);

        return mapToDTO(savedApp);
    }

    /**
     * Start a blocking session
     */
    @Transactional
    public BlockingSessionDTO startBlockingSession(String userId, StartSessionRequest request) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        int duration = request.getDuration() != null ? request.getDuration() : DEFAULT_SESSION_DURATION;

        BlockingSession session = new BlockingSession(user, request.getAppId(), request.getAppName(), LocalDateTime.now(), duration, BlockingSession.SessionStatus.ACTIVE, 0, 0);

        BlockingSession savedSession = blockingSessionRepository.save(session);

        // Log the action
        logBlockingAction(user, request.getAppId(), request.getAppName(),
                BlockingLog.LogAction.SESSION_START, 0, savedSession.getId());

        return mapToSessionDTO(savedSession);
    }

    /**
     * Complete a blocking session (reward user)
     */
    @Transactional
    public SessionActionResponse completeBlockingSession(String userId, String sessionId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        BlockingSession session = blockingSessionRepository.findById(sessionId)
                .orElseThrow(() -> new ResourceNotFoundException("Session not found"));

        if (!session.getUser().getUserId().equals(userId)) {
            throw new UnauthorizedActionException("You do not have access to this session");
        }

        if (session.getStatus() != BlockingSession.SessionStatus.ACTIVE) {
            throw new BadRequestException("Session is not active");
        }

        // Update session
        session.setStatus(BlockingSession.SessionStatus.COMPLETED);
        session.setCompletedAt(LocalDateTime.now());
        int rewardCoins = calculateRewardCoins(1);
        session.setCoinsEarned(rewardCoins);
        blockingSessionRepository.save(session);

        // Award coins
        int currentCoins = updateUserCoins(userId, rewardCoins);

        // Log the action
        logBlockingAction(user, session.getAppId(), session.getAppName(),
                BlockingLog.LogAction.SESSION_COMPLETE, rewardCoins, sessionId);

        return new SessionActionResponse(sessionId, rewardCoins, currentCoins, "Session completed! Earned " + rewardCoins + " Focus Points");
    }

    /**
     * Break a blocking session (penalize user)
     */
    @Transactional
    public SessionActionResponse breakBlockingSession(String userId, String sessionId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        BlockingSession session = blockingSessionRepository.findById(sessionId)
                .orElseThrow(() -> new ResourceNotFoundException("Session not found"));

        if (!session.getUser().getUserId().equals(userId)) {
            throw new UnauthorizedActionException("You do not have access to this session");
        }

        if (session.getStatus() != BlockingSession.SessionStatus.ACTIVE) {
            throw new BadRequestException("Session is not active");
        }

        // Update session
        session.setStatus(BlockingSession.SessionStatus.BROKEN);
        session.setBrokenAt(LocalDateTime.now());
        session.setCoinsLost(PENALTY_COINS);
        blockingSessionRepository.save(session);

        // Deduct coins
        int currentCoins = updateUserCoins(userId, -PENALTY_COINS);

        // Log the action
        logBlockingAction(user, session.getAppId(), session.getAppName(),
                BlockingLog.LogAction.SESSION_BROKEN, -PENALTY_COINS, sessionId);

        return new SessionActionResponse(sessionId, -PENALTY_COINS, currentCoins, "Session broken! Lost " + PENALTY_COINS + " Focus Points");
    }

    /**
     * Get active blocking sessions
     */
    public List<BlockingSessionDTO> getActiveBlockingSessions(String userId) {
        List<BlockingSession> sessions = blockingSessionRepository
                .findByUser_UserIdAndStatus(userId, BlockingSession.SessionStatus.ACTIVE);

        return sessions.stream()
                .map(this::mapToSessionDTO)
                .collect(Collectors.toList());
    }

    /**
     * Get all blocking sessions
     */
    public List<BlockingSessionDTO> getAllBlockingSessions(String userId) {
        List<BlockingSession> sessions = blockingSessionRepository.findByUser_UserId(userId);

        return sessions.stream()
                .map(this::mapToSessionDTO)
                .collect(Collectors.toList());
    }

    /**
     * Get blocking statistics
     */
    public BlockingStatsDTO getBlockingStats(String userId) {
        List<BlockingSession> sessions = blockingSessionRepository.findByUser_UserId(userId);

        int totalSessions = sessions.size();
        int completedSessions = 0;
        int brokenSessions = 0;
        int totalCoinsEarned = 0;
        int totalCoinsLost = 0;

        for (BlockingSession session : sessions) {
            if (session.getStatus() == BlockingSession.SessionStatus.COMPLETED) {
                completedSessions++;
                totalCoinsEarned += session.getCoinsEarned();
            } else if (session.getStatus() == BlockingSession.SessionStatus.BROKEN) {
                brokenSessions++;
                totalCoinsLost += session.getCoinsLost();
            }
        }

        double successRate = totalSessions > 0 ? (completedSessions * 100.0 / totalSessions) : 0;

        return new BlockingStatsDTO(totalSessions, completedSessions, brokenSessions, totalCoinsEarned, totalCoinsLost, successRate);
    }

    /**
     * Cleanup stale sessions (for debugging)
     */
    @Transactional
    public int cleanupStaleSessions(String userId) {
        List<BlockingSession> activeSessions = blockingSessionRepository
                .findByUser_UserIdAndStatus(userId, BlockingSession.SessionStatus.ACTIVE);

        for (BlockingSession session : activeSessions) {
            session.setStatus(BlockingSession.SessionStatus.BROKEN);
            session.setBrokenAt(LocalDateTime.now());
            session.setCoinsLost(0); // No penalty for cleanup
        }

        blockingSessionRepository.saveAll(activeSessions);
        return activeSessions.size();
    }

    // Helper methods

    private int updateUserCoins(String userId, int coinsChange) {
        UserStats stats = userStatsRepository.findByUserId(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User stats not found"));

        long currentCoins = stats.getCurrentFocusPoints();
        long newCoins = Math.max(0, currentCoins + coinsChange);

        if (coinsChange > 0) {
            stats.setTotalFocusPoints(stats.getTotalFocusPoints() + coinsChange);
        }
        stats.setCurrentFocusPoints(newCoins);
        userStatsRepository.save(stats);

        return (int) newCoins;
    }

    private int calculateRewardCoins(int blockedAppCount) {
        return Math.max(0, blockedAppCount) * REWARD_COINS_PER_BLOCKED_APP;
    }

    private void logBlockingAction(User user, Integer appId, String appName,
                                    BlockingLog.LogAction action, int coinsChange, String sessionId) {
        BlockingLog log = new BlockingLog(user, appId, appName, action, coinsChange, sessionId);
        blockingLogRepository.save(log);
    }

    private BlockedAppDTO mapToDTO(BlockedApp app) {
        return new BlockedAppDTO(app.getId(), app.getAppId(), app.getAppName(), app.getCategory(), app.getIcon(), app.getLogo(), app.getBlocked(), app.getPackageName(), app.getBundleId(), app.getUpdatedAt());
    }

    private BlockingSessionDTO mapToSessionDTO(BlockingSession session) {
        return new BlockingSessionDTO(session.getId(), session.getAppId(), session.getAppName(), session.getStartTime(), session.getDuration(), session.getStatus().name().toLowerCase(), session.getCoinsEarned(), session.getCoinsLost(), session.getCompletedAt(), session.getBrokenAt(), session.getCreatedAt());
    }
}
